frame/staking/src/tests.rs

```diff
@@ -4278,6 +4282,76 @@ mod election_data_provider {
 			assert_eq!(ForceEra::<Test>::get(), Forcing::NotForcing);
 		})
 	}
+
+	#[test]
+	fn can_paginate() {
+		ExtBuilder::default()
+			.add_staker(61, 60, 500, StakerStatus::Nominator(vec![11]))
+			.add_staker(71, 70, 500, StakerStatus::Nominator(vec![11]))
+			.add_staker(81, 80, 500, StakerStatus::Nominator(vec![11]))
+			.add_staker(91, 90, 500, StakerStatus::Nominator(vec![11]))
+			.set_status(31, StakerStatus::Idle)
+			.set_status(41, StakerStatus::Idle)
+			.build_and_execute(|| {
+				// we shall have 2 validators and 5 nominators.
+				assert_eq!(Nominators::<Test>::count(), 5);
+				assert_eq!(Validators::<Test>::count(), 2);
+
+				// a correct, 2 page snapshot.
+				Pages::set(2);
+				// we have 6 voters in total, get them over two pages, each 4 max.
+				assert_eq!(
+					Staking::electing_voters_paged(Some(4), 1).unwrap(),
+					vec![
+						// everything sorted based on stake.
+						(11, 1000, bounded_vec![11]),
+						(21, 1000, bounded_vec![21]),
+						(101, 500, bounded_vec![11, 21]),
+						(61, 500, bounded_vec![11]),
+					]
+				);
+
+				assert_eq!(LastIteratedNominator::<Test>::get(), Some(61));
+
+				// tandem: node 61 cannot be chilled in any humanly possible way now.
+				let _ = with_transaction(|| -> TransactionOutcome<DispatchResult> {
+					// normal chilling
+					assert_noop!(
+						Staking::chill(Origin::signed(60)),
+						Error::<Test>::TemporarilyNotAllowed
+					);
+					// chill-other with the virtue of being less than min-bond
+					MinNominatorBond::<Test>::put(501);
+					ChillThreshold::<Test>::put(Percent::default());
+					MaxNominatorsCount::<Test>::put(1);
+					assert_noop!(
+						Staking::chill_other(Origin::signed(100), 60),
+						Error::<Test>::TemporarilyNotAllowed
+					);
+					storage::TransactionOutcome::Rollback(Ok(()))
+				});
+
+				// and making this nominator un-decodable.
+				MaxNominations::set(0);
+				assert_noop!(
+					Staking::chill_other(Origin::signed(100), 60),
+					Error::<Test>::TemporarilyNotAllowed
+				);
+				// this needs to reverted, despite being transactional scope, because it is
+				// thread-local, not storage.
```

Comment:
Interesting - I thought with the thread local stuff it should have no side effects in other tests

Reply:
Its also not in the transactional scope I think - just the outermost scope of the closure given to `build_and_execute`

Reply:
Its also not in the transactional scope I think - just the outermost scope of the closure given to `build_and_execute`